### PR TITLE
fix: return query if it already exists

### PR DESCRIPTION
**Pull request by eschutho**
### SUMMARY
We were seeing some duplicate client_id errors when running queries. Here's the root cause: 

When we have a timeout on a query, we return a 408 , and when most browsers get a 408 response, they automatically retry the request. The issue with the query POST request is that the db request happens after we create the query, so if the db request times out, then the retry fails because the request is not idempotent, due to the fact that the client_id is created on the client and has a uniqueness constraint in the db. 

My solution here is to check to see if the query exists and if it is in a pending or timed out state, don't resave it, but rather just return the original query.

I had thought of a few different options for this issue, and here are the reasonings why I didn't go with other solutions: 
1) change client_id to query.id and generate it server side. The reason that we have the id generated client side and stored in the db is so that we can stop the query before the original response has returned. 
2) return something other than a 408 when the db has timed out. If the db timeout is long, it's possible that the server will time out before the db and we'll likely still be left with the same issue. 
3) run the db request before saving the query. Because the same endpoint is used with async query requests, the refactor would be much more prone to errors. 

Some concerns/questions: 
If a query is new, and happens to match an existing query that is running, (unlikely but possible), we don't want to return someone else's query. There's a check on this lookup for user_id, but that can be null. There's also a lookup on sql_editor_id. My assumption is that with the combination of these three factors, we should with certainty only get the request that we are intending, but I'd welcome another set of eyes on that. The last option is to check the sql string, but the comparison on what could potentially be a long string may be costly. 

In my testing, I was never able to reproduce a server timeout response, only a netstat error, and the query continued to have a status of "running" which means that it will continue to be polled for. You have to explicitly delete the query if you navigate away from the page without stopping the query. I didn't change that behavior, but it's something that we may want to investigate. 

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
No visual changes

### TESTING INSTRUCTIONS
If you are testing locally and can put a debug statement somewhere in the sql_json POST request you can replicate a timeout. After about 2 minutes, you should see another sql_json POST request in your browser network tab, and it should return a 200 with the original query, and not raise an error. 

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/views/sql_lab.py [hunk 1]
```diff
@@ -228,10 +230,12 @@ def migrate_query(  # pylint: disable=no-self-use
...
     @expose("<int:tab_state_id>/query/<client_id>", methods=["DELETE"])
     def delete_query(  # pylint: disable=no-self-use
-        self, tab_state_id: str, client_id: str
+        self, tab_state_id: int, client_id: str
     ) -> FlaskResponse:
         db.session.query(Query).filter_by(
-            client_id=client_id, user_id=g.user.get_id(), sql_editor_id=tab_state_id
+            client_id=client_id,
+            user_id=g.user.get_id(),
+            sql_editor_id=str(tab_state_id),
```
Comment (eschutho) [inline superset/views/sql_lab.py:238]: I was getting an error when trying to delete the query b/c it was passing in an int here.

**Review by betodealmeida — APPROVED** [review]
This looks good... such a tricky bug! I wasn't familiar with 408.

**Comment by eschutho** [comment]
did a visual test  on ephemeral env and it looked good.

**Comment by henryyeh** [comment]
🏷️ 2021.25

**Pull request merged**